pkg/crosscluster/logical/table_batch_handler.go

```diff
@@ -0,0 +1,239 @@
+	// kvLwwLosers is the number of rows that were dropped as lww losers after
+	// attempting to write to the KV layer. This case only occurs if there is a
+	// tombstone that is more recent than the replicated action.
+	kvLwwLosers int64
+}
+
+func (t *tableBatchStats) Add(o tableBatchStats) {
+	t.refreshedRows += o.refreshedRows
+	t.inserts += o.inserts
+	t.updates += o.updates
+	t.deletes += o.deletes
+	t.tombstoneUpdates += o.tombstoneUpdates
+	t.refreshLwwLosers += o.refreshLwwLosers
+	t.kvLwwLosers += o.kvLwwLosers
+}
+
+// newTableHandler creates a new tableHandler for the given table descriptor ID.
+// It internally constructs the sqlReader and sqlWriter components.
+func newTableHandler(
+	tableID descpb.ID,
+	db descs.DB,
+	codec keys.SQLCodec,
+	sd *sessiondata.SessionData,
+	leaseMgr *lease.Manager,
+	settings *cluster.Settings,
+) (*tableHandler, error) {
+	var table catalog.TableDescriptor
+
+	// NOTE: we don't hold a lease on the table descriptor, but validation
+	// prevents users from changing the set of columns or the primary key of an
+	// LDR replicated table.
+	err := db.DescsTxn(context.Background(), func(ctx context.Context, txn descs.Txn) error {
+		var err error
+		table, err = txn.Descriptors().GetLeasedImmutableTableByID(ctx, txn.KV(), tableID)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	reader, err := newSQLRowReader(table)
+	if err != nil {
+		return nil, err
+	}
+
+	writer, err := newSQLRowWriter(table)
+	if err != nil {
+		return nil, err
+	}
+
+	tombstoneUpdater := newTombstoneUpdater(codec, db.KV(), leaseMgr, tableID, sd, settings)
+
+	return &tableHandler{
+		sqlReader:        reader,
+		sqlWriter:        writer,
+		db:               db,
+		tombstoneUpdater: tombstoneUpdater,
+	}, nil
+}
+
+func (t *tableHandler) handleDecodedBatch(
+	ctx context.Context, batch []decodedEvent,
+) (tableBatchStats, error) {
+	stats, err := t.attemptBatch(ctx, batch)
+	if err == nil {
+		return stats, nil
+	}
+
+	refreshedBatch, refreshStats, err := t.refreshPrevRows(ctx, batch)
+	if err != nil {
+		return stats, err
+	}
+
+	stats, err = t.attemptBatch(ctx, refreshedBatch)
+	if err != nil {
+		return tableBatchStats{}, err
+	}
+
+	stats.Add(refreshStats)
+
+	return stats, nil
+}
+
+func (t *tableHandler) attemptBatch(
+	ctx context.Context, batch []decodedEvent,
+) (tableBatchStats, error) {
+	var stats tableBatchStats
+	err := t.db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		for _, event := range batch {
+			switch {
+			case event.isDelete && len(event.prevRow) != 0:
+				stats.deletes++
+				err := t.sqlWriter.DeleteRow(ctx, txn, event.originTimestamp, event.prevRow)
+				if isLwwLoser(err) {
+					stats.kvLwwLosers++
+					continue
+				}
+				if err != nil {
+					return err
+				}
+			case event.isDelete && len(event.prevRow) == 0:
+				stats.tombstoneUpdates++
+				tombstoneUpdateStats, err := t.tombstoneUpdater.updateTombstone(ctx, txn, event.originTimestamp, event.row)
```

Comment:
it may be easier on the eyes for the caller of updateTombstone to handle LWW and update stats.

Reply:
Yeah, i'm fine keeping this as is.